step3 - 라자냐

wonhee009 opened the pull request.

jryoun1 committed:
version차이로 인한 수정

jryoun1 committed:
feat: UserInformation 타입 생성

jryoun1 committed:
project 설정 변경

jryoun1 committed:
Merge pull request #1 from jryoun1/step1-jake

회원가입 화면 프로젝트 #step1 lasagna, jake

jryoun1 committed:
refactor: 변수명 수정 및 타입 변경

jryoun1 committed:
error

wonhee009 committed:
fix

wonhee009 committed:
fix storyboard

jryoun1 committed:
feat: 새로운 VC 추가

jryoun1 committed:
feat: 새로운 VC 추가

wonhee009 committed:
Merge branch 'step2-jake' into main

jryoun1 committed:
fix: conflict merge bug 수정

wonhee009 committed:
feat: add IBOutlet

wonhee009 committed:
feat: set keyboard secure

wonhee009 committed:
feat: edit ID textfield return key style

wonhee009 committed:
feat: add SingUpVC

wonhee009 committed:
feat: move SIngUpVC

wonhee009 committed:
feat: add hide keyboard

wonhee009 committed:
feat: edit present style

wonhee009 committed:
feat: edit method location

wonhee009 committed:
feat: edit setting

wonhee009 committed:
fix: edit name

wonhee009 committed:
Merge branch 'step2-lasagna' into main

jryoun1 committed:
storyboard 구현

wonhee009 committed:
feat: storyboard

jryoun1 committed:
feat: step3 과정1 view 생성

wonhee009 committed:
feat: edit IBOutlet name

wonhee009 committed:
refactor: edit ViewController name

wonhee009 committed:
refactor: edit gesture recognizer

wonhee009 committed:
refactor: edit textField name

wonhee009 committed:
Merge branch 'step3-jake' into step2-lasagna

wonhee009 committed:
Merge branch 'step2-lasagna' into main

wonhee009 committed:
feat: add textField placeholder

wonhee009 committed:
feat: button check unEnable

wonhee009 committed:
feat: add photo, album privacy message

wonhee009 committed:
feat: add imagePicker

wonhee009 committed:
feat: add Groups

wonhee009 committed:
feat: add open Album

wonhee009 committed:
feat: add privacy info

wonhee009 committed:
feat: edit crop image

wonhee009 committed:
feat: add textView outlet

wonhee009 committed:
feat: add textview placeholder

wonhee009 committed:
feat: add textview placeholder logic

wonhee009 committed:
feat: edit return key type

wonhee009 committed:
feat: textview add accesory view

wonhee009 committed:
feat: add textview border

wonhee009 committed:
feat: add accesoryView, return next button

wonhee009 committed:
feat: add moveNextField

wonhee009 committed:
feat: adjust dark mode textcolor

wonhee009 committed:
feat: edit textView trimming func

wonhee009 committed:
feat: add mark

wonhee009 committed:
feat: add signUpOptionViewcontroller

wonhee009 committed:
feat: add error

wonhee009 committed:
feat: button flow

wonhee009 committed:
feat: setting imageView contentMode

wonhee009 committed:
refactor: add New Group

yagom commented:
스텝 반영사항이나 고민거리에 대한 코멘트가 없는데, 아직 PR요청을 하지 않는 상태인건가요? :) 
아직 커밋할 거리가 남아있나요?

wonhee009 commented:
@yagom 아! 야곰 코멘트를 까먹었어요😂
- textField에서 next를 누르면 다음 field로 이동하도록 하기 위해 IBOutlet Collection으로 구현했습니다.
- 키보드에 액세서리 뷰를 추가해서 완료 버튼을 클릭시 키보드를 hide했습니다.
- textView에는 placeholder가 없어서 delegate로 placeholder처럼 구현했습니다.
- textField들을 tag로 관리하려했고 tag 숫자를 enum을 통해 type으로 관리했습니다.

Limwin94 commented on SignUpFlow/SignUpFlow/Assets.xcassets/image_asset/yagom.imageset/Contents.json:
```diff
@@ -0,0 +1,21 @@
+{
+  "images" : [
+    {
+      "filename" : "yagom.png",
```
사진이 1x 만 있는 이유가 따로 있을까요?
2x, 3x는 무엇을 의미하는 걸까요?

wonhee009 replied:
디바이스 해상도에 따라 이미지를 적용하기 위해 이미지 배율을 구별해놨습니다.
해당 프로젝트에서는 이미지가 중요하다 생각하지 않아서 x1 배율로만 넣어놨습니다.

Limwin94 commented on SignUpFlow/SignUpFlow/Extension/UIViewControllerExtension.swift:
```diff
@@ -0,0 +1,26 @@
+
+extension UIViewController {
+    func errorAlert(_ error: Error) -> UIAlertController {
+        var errorMessgae: String?
```
오타 수정바랍니다 :)

Limwin94 commented on SignUpFlow/SignUpFlow/SignUpError.swift:
```diff
@@ -0,0 +1,31 @@
+    case unknown
+}
+
+extension SignUpError : LocalizedError {
```
어떤 에러 내용인지 따로 커스텀해서 사용하셨네요.
사용자 입장에서 왜 안되는지 알 수 있어서 좋은 것 같습니다 👍🏻

Limwin94 commented on SignUpFlow/SignUpFlow/SignUpError.swift:
```diff
@@ -0,0 +1,31 @@
+
+enum SignUpError : Error {
+    case getText
+    case isNotValidText
+    case isNotValidEmail
```
invalid대신 isNot을 사용한 이유가 있을까요?

wonhee009 replied:
제가.... 영어가 짧아서... 생각을 못했네요😂 이번 기회에 다음 프로젝트에서 invalid 써먹을 수 있을거 같아요!ㅋㅋ🙌

Limwin94 commented on SignUpFlow/SignUpFlow/Extension/UIColorExtension.swift:
```diff
@@ -0,0 +1,25 @@
+
+extension UIColor {
+    static var textViewColor: UIColor {
+        if #available(iOS 13, *) {
```
textView에 대해서 다크 모드를 지원했네요 👍🏻 
찾아보니 textView의 textColor에 대해서 적용해주는데, textViewColor란 네이밍은 textView 자체에 관련된 Color를 설정해주는 의미로 다가오는것 같습니다.

Limwin94 commented on SignUpFlow/SignUpFlow/SignUpViewController.swift:
```diff
@@ -0,0 +1,219 @@
+    
+    func textViewDidEndEditing(_ textView: UITextView) {
+        guard let introductionText = textView.text else {
+            return self.showError(SignUpError.getText)
```
textView의 text를 언제 가져오지 못할까요?

yagom replied:
textView의 text가 `nil`인 경우와 `""`인 경우에 이 guard의 동작은 어떻게 달라질까요?
`nil`과 `""`의 차이는 무엇일까요?

wonhee009 replied:
@Limwin94 @yagom 
우선 guard let 을 썼던 이유는 textView.text가 옵셔널여서 바인딩하고 싶어서 사용했습니다.
그 과정에서 당연히 nil이 들어오는상황이 발생할 수 있다 생각해서 사용했었습니다.

textView.text에 대해서 공식 문서에서 default로 `""`를 반환한다 나와 있어 그렇다면 왜 굳이 옵셔널로 return해줄까라는 생각에 알아봤고 Objectivew-C와의 호환성으로 optionl로 return된다는 사실을 알 수 있었고,
Swift에서는 textView.text = nil이 textView.text = ""이 된다는 사실을 알 수 있었습니다.
![스크린샷 2020-12-04 11 51 32](https://user-images.githubusercontent.com/30456206/101116478-87363300-3628-11eb-8f18-481aac2bee11.png)

guard보다 if let이 좀 더 맞지 않았을까하는 생각이 지금은 좀 드는거 같아요!ㅋㅋ

그리고 야곰의 리뷰에 대한 답변으로는
우선 제 코드에서 nil과 ""인 경우 동작 차이를 물어보셨는데
`nil`이라면 아래 text의 유효성을 검사하는 코드를 돌지 않고, `""`이라면 아래의 코드를 돌았겠네요.

nil은 정말 아무것도 없는 것 ""은 문자열이 비어있지만 존재하는 문자열이죠!

yagom replied:
문서에서 ""를 반환한다고 하더라도 타입 자체가 옵셔널이기 때문에 nil이 들어올 수 있음을 항상 염두해야 함은 맞지요.
그러나 코드에서는 두 경우의 수를 모두 제어하기엔 어려움이 있어보여 질문했습니다 :)

wonhee009 replied:
guard let으로 nil을 제어하고, 밑에 isValidateText(from:)에서 ""값을 제어하고 있었는데 제가 놓치고 있는게 있었을까요?

Limwin94 commented on SignUpFlow/SignUpFlow/SignUpViewController.swift:
```diff
@@ -0,0 +1,219 @@
+    private func moveNextTextField(from textField: UITextField) {
+        let nextTag = textField.tag + 1
+        if nextTag <= self.profileTextFields.count {
+            let field = profileTextFields.first(where: { $0.tag == nextTag })
```
IBOutlet Collection + tag를 잘 활용한것 같습니다 👍🏻

Limwin94 commented on SignUpFlow/SignUpFlow/Data/UserInformation.swift:
```diff
@@ -0,0 +1,33 @@
+import Foundation
+import UIKit
+
+final class UserInformation {
```
UserInformation과 UserInfo가 하는 역할이 다르기 때문에 나누어 놓은건가요? (궁금🧐)

wonhee009 replied:
UserInformation은 사용자 정보를 가지고 있는 data로 사용하기 위해 작성했고, UserInfo는 회원가입 과정에서 각 스텝별로 정보를 관리하기 위해 분리해뒀습니다.

Limwin94 reviewed (commented):
고생하셨습니다. 

코드내에서 사용자 편의성, 많이 고민한 부분이 느껴졌습니다.
코멘트 확인 부탁드립니다.

yagom commented on SignUpFlow/SignUpFlow/Data/UserInfo.swift:
```diff
@@ -0,0 +1,42 @@
+import Foundation
+import UIKit
+
+class UserInfo : InfoForm {
```
다른 클래스와 다르게 이 타입만 띄어쓰기 스타일이 다른 이유가 있을까요?

yagom replied:
타입의 이름도 다른 이름들과 마찬가지로 축약형은 피하고 있습니다.
UserInformation의 축약형으로 생각하는데 어떻게 생각하는지요?

wonhee009 replied:
@yagom
축약형이라고 생각해요!
축약형을 피하려고 노력은 하는데 가끔 놓치는 부분이 생기는거 같아요!
그리고 UserInfo라는 네이밍이 제가 의도한 네이밍과는 조금 거리가 있는거 같아서 네이밍을 수정했습니다.

yagom commented on SignUpFlow/SignUpFlow/Data/UserInfo.swift:
```diff
@@ -0,0 +1,42 @@
+
+class UserInfo : InfoForm {
+    var id: String?
+    var checkPassword: String?
```
check가 password 앞에 있느냐, 뒤에 있느냐에 따라 동사가 될지말지가 달라지는 것 같아요 :) 지금의 형태라면 동사형이라 메서드 이름에 더 어울리지 않을까요?

yagom replied:
요소의 이름을 바꿀 때는 Xcode의 Refactor - Rename 기능을 꼭 사용해볼 것을 추천합니다.

yagom commented on SignUpFlow/SignUpFlow/Data/UserInfo.swift:
```diff
@@ -0,0 +1,42 @@
+    var image: UIImage?
+    var introduction: String?
+    
+    func isFilled() -> Bool {
```
이렇게 메서드로 구현하는 것과 연산 프로퍼티로 구현하는 것에는 어떤 차이가 있을까요?
언제 연산 프로퍼티를 사용하고 언제 메서드를 사용할지, 그 둘을 사용하는 기준이 있을까요?

wonhee009 replied:
@yagom
전에 붱이랑 이 부분에 대해서 얘기 했었는데 메서드는 property의 값을 변경하는 작업이 있다면 메서드를 사용하고 없다면 프로퍼티를 사용한다고 얘기했어서 그렇게 기준을 세우고 사용하려 하고 있었습니다!
이 부분은 연산 프로퍼티로 변경해서 커밋했습니다!

yagom commented on SignUpFlow/SignUpFlow/Data/UserInformation.swift:
```diff
@@ -0,0 +1,33 @@
+    var birthDate: Date?
+}
+
+protocol UserInfoForm {
+    func isFilled() -> Bool
+    func clearInfo()
+    func validateInfo() throws -> Bool
+}
+
+protocol InfoForm: UserInfoForm {
+    var id: String? { get }
+    var password: String? { get }
+    var checkPassword: String? { get }
+    var image: UIImage? { get }
+    var introduction: String? { get }
+}
```
프로토콜 이름만으로 무엇을 요구하게 될지 전혀 예측이 안되는 것 같아요. 그리고 프로토콜을 따로 구현한 것이 아니라 상속한 이유도 궁금합니다. 
애초에 프로토콜을 사용하기로 했던 계기나 이유도 궁금합니다 :)

yagom commented on SignUpFlow/SignUpFlow/MainViewController.swift:
```diff
@@ -0,0 +1,31 @@
+    @IBOutlet weak var idTextField: UITextField!
+    @IBOutlet weak var passwordTextField: UITextField!
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+    }
```
사용하지 않는 메서드를 남겨둘 필요가 있을까요?

wonhee009 replied:
이후 스텝 진행하면서 main view textField들의 초기화할때 사용하려 남겨두었습니다.
(next 버튼 누르면 다음으로 넘어가고 그런 부분을 추가로 구현하고싶어서요!)

yagom replied:
이후에 사용 가능성이 있더라도 커밋 순간에는 사용하지 않는 코드는 남겨두지 않는 것이 좋겠습니다 :)

yagom commented on SignUpFlow/SignUpFlow/MainViewController.swift:
```diff
@@ -0,0 +1,31 @@
+        self.view.endEditing(true)
+    }
+    
+    @IBAction func tapSignupButton() {
```
touchUp, touchDown, tap 등의 차이는 뭘까요? 
메서드의 이름을 모두 tap으로 시작하도록 처음부터 약속된건가요?
만약 그렇다면 그렇게 약속한 이유가 있나요?

yagom commented on SignUpFlow/SignUpFlow/SignUpViewController.swift:
```diff
@@ -0,0 +1,219 @@
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        setKeyboard()
+        setProfileImage()
+        setProfileTextField()
+        setIntroductionTextView()
```
set은 그것 자체를 넣어주는 이름인 것 같은데...
configure, setUp, installation 등의 이름을 배제한 이유가 있을까요?

yagom commented on SignUpFlow/SignUpFlow/SignUpOptionViewController.swift:
```diff
@@ -0,0 +1,19 @@
+
+class SignUpOptionViewController: UIViewController {
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+    }
```
check

wonhee009 replied:
이 부분도 이후 스텝을 진행하면서 사용할 부분이라 남겨두었습니다!

yagom commented on SignUpFlow/SignUpFlow/SignUpViewController.swift:
```diff
@@ -0,0 +1,219 @@
+        setIntroductionTextView()
+    }
+    
+    override func viewWillAppear(_ animated: Bool) {
```
super 메서드를 호출하지 않은 이유가 있나요?

wonhee009 replied:
까먹었슴다...ㅋㅋ 다음부터는 좀 더 꼼꼼하게 해볼게요!😂

yagom commented on SignUpFlow/SignUpFlow/SignUpViewController.swift:
```diff
@@ -0,0 +1,219 @@
+    }
+    
+    // MARK: - settings
+    private func setKeyboard() {
```
키보드 액세서리 뷰를 사용하다니, 사용자가 훨씬 편리하게 키보드를 감출 수 있겠네요 👍

yagom commented on SignUpFlow/SignUpFlow/SignUpViewController.swift:
```diff
@@ -0,0 +1,219 @@
+    }
+    
+    // MARK: - Step Buttons
+    @IBAction func tapCancleButton(_ sender: Any) {
```
typo

yagom commented on SignUpFlow/SignUpFlow/SignUpViewController.swift:
```diff
@@ -0,0 +1,219 @@
+}
+
+extension SignUpViewController : UITextViewDelegate {
+    func setIntroductionPlaceholder() {
```
이 메서드가 이 extension과 묶이게 된 이유가 있을까요?

wonhee009 replied:
textView에 placeholder가 없어서 트릭?으로 placeholder처럼 보이도록 구현했습니다.
 textViewDidEndEditing(_:)에서 textView의 text가 빈 문자열일 경우 placeholder처럼 보이게 바꿔치기하기 위해 작성한 func였고, textViewDidEndEditing(_:)에서 호출하려 했었는데 그 부분을 그냥 코드로 짜버렸네요 
그 부분 메서드로 대체해서 커밋해뒀습니다!
extension에 묶인 이유는 해당 메서드가 UITextViewDelegate에 더 가깝다 생각해서였씁니다.

yagom reviewed (commented):
몇 가지 조금 더 첨언합니다 :)

wonhee009 committed:
fix: edit typo

wonhee009 committed:
refactor: edit property name

기존의 네이밍이 textViewColor를 설정해주는 의미로 다가온다는 리뷰에 textView의 textColor임을 알 수 있도록 수정했습니다.

wonhee009 committed:
refactor: guard let -> if let 

textView, textField의 text를 확인하는 과정에서 guard let에서 if let으로 바인딩을 변경하였습니다.

wonhee009 committed:
refactor: edit class naming

wonhee009 committed:
refactor: typo

wonhee009 committed:
refactor: tapCancleButton

취소 버튼 클릭시 관련 정보 clear

wonhee009 committed:
refactor: add setting UI func

wonhee009 committed:
refactor: property naming

wonhee009 committed:
refactor: mothod to property

wonhee009 committed:
refactor: delete additional code

wonhee009 committed:
refactor: replace code block to method

yagom commented:
추가 리뷰를 희망하면 모든 커밋을 보내고 리뷰어를 태그해주세요 :)

wonhee009 committed:
refactor: class naming

wonhee009 committed:
refactor: edit touchUp Method naming

wonhee009 commented:
@yagom 
프로토콜의 이름이 요구사항이 반영되지 않는다는 의견에 동의하여 SignUpForm으로 변경해 의도를 좀 더 명확하게 드러내려 노력했습니다.
프로토콜을 사용했던 이유는 각 스텝마다 가지고 있는 정보가 다르고 체크해야하는 정보도 다르다고 판단해서 그걸 제약?해주고 싶어 사용했습니다.
상속했던 이유는 해당 스텝과 다음 스텝에서 form이 채워져 있는지 확인하고 데이터를 지우는 func들이 공통으로 사용되야한다 생각했습니다.

tap이나 click으로 시작하는 메서드명으로 네이밍했었는데 다른 캠퍼들과 의견을 나눈 결과 제스처를 명확하게 표시하기 위해 앞에 touchUp으로 시작하는 네이밍으로 변경햇습니다.

configure, setUp, installation 등을 배제한 것은 아녔어요! 
set이 설정해준다는 느낌이 저에게 있어서 set으로 네이밍을 했는데 야곰이 써준 키워드들의 뜻을 살펴보고 setUp이 더 적절하다 생각해 바꿔서 커밋했습니다!

AppleCEO commented on SignUpFlow/SignUpFlow/Data/SignUpForm.swift:
```diff
@@ -0,0 +1,42 @@
+    var introduction: String?
+    
+    var isFilled: Bool {
+        guard let _ = self.image,
+              let _ = self.id,
+              let validatePassword = self.password,
+              let validateCheckPassword = self.passwordCheck,
+              let _ = self.introduction,
+              validatePassword == validateCheckPassword
```
guard let에 와일드 카드 패턴까지 고급스러운 코드입니다!
와일드 카드 패턴을 하지 말고 다른 방식으로 값이 있는지 확인할 수 있을까요?

AppleCEO commented on SignUpFlow/SignUpFlow/Info.plist:
```diff
@@ -20,6 +20,10 @@
 	<string>1</string>
 	<key>LSRequiresIPhoneOS</key>
 	<true/>
+	<key>NSCameraUsageDescription</key>
+	<string>프로필 사진을 가져오기 위해 카메라를 사용합니다.</string>
+	<key>NSPhotoLibraryUsageDescription</key>
+	<string>프로필 사진을 가져오기 위해 앨범을 사용합니다.</string>
```
설명을 명확히 적은 것은 아주 좋습니다!
이 설명이 불명확하면 앱스토어에 못올리는 경우도 있습니다~

AppleCEO commented on SignUpFlow/SignUpFlow/SignUpViewController.swift:
```diff
@@ -0,0 +1,226 @@
+    @IBOutlet weak var nextButton: UIButton!
+    
+    private let profileImagePicker = UIImagePickerController()
+    private let introductionPlaceholderMessage = "자기소개를 입력해주세요."
```
문구를 이런 식으로 상수에 저장하는 것은 아주 좋은 습관입니다 💯

AppleCEO commented on SignUpFlow/SignUpFlow/SignUpError.swift:
```diff
@@ -0,0 +1,31 @@
+
+import Foundation
+
+enum SignUpError : Error {
+    case getText
+    case isNotValidText
+    case isNotValidEmail
```
에러들을 enum에 모아놓은 것은 아주 좋은 습관입니다!

AppleCEO commented on SignUpFlow/SignUpFlow/SignUpError.swift:
```diff
@@ -0,0 +1,31 @@
+}
+
+extension SignUpError : LocalizedError {
+    var errorDescription: String? {
+        switch self {
+        case .getText:
+            return "텍스트를 가져오는데 문제가 있습니다.\n잠시 후 다시 시도해주세요."
+        case .isNotValidText:
+            return "유효하지 않은 텍스트입니다.\n올바르게 입력해주세요."
+        case .isNotValidEmail:
+            return "유효하지 않은 이메일 주소입니다.\n올바르게 입력해주세요."
+        case .unknown:
+            return "알 수 없는 오류가 발생했습니다.\n잠시 후 다시 시도해주세요."
+        }
+    }
```
열거형에 프로퍼티를 저장해보셨군요! 
이렇게 만드는 것과 열거형에 rawValue로 문자열을 저장하는 것과 어떤 차이가 있을까요?

AppleCEO reviewed (commented):
전체적으로 완성도가 높네요!
열거형을 적극적으로 활용한 부분도 너무 좋았습니다.
몇가지 코멘트 남겨놨습니다.
확인해주시면 다음 단계로 넘어가도록 merge해드리겠습니다~

wonhee009 commented:
@AppleCEO 머지 부탁드립니다🙇‍♂️

AppleCEO commented:
> @AppleCEO 머지 부탁드립니다🙇‍♂️

코멘트에 대해 피드백 해주시면 확인해보고 머지 드리겠습니다! 🥇

wonhee009 commented:
@AppleCEO 

- guard let을 썼던 이유는 옵셔널 바인딩을 위해 썼었습니다. 와일드 카드 패턴을 이용하지 않고 확인하는 방법으로는`guard self.image != nil`이 생각나는데 이 부분을 말씀하신걸까요?

- Error에 관한 enum여서 Error 메시지는 반드시 설정해야 한다고 생각했습니다. rawValue에 지정하게 되면 Error가 추가되는 상황에서 메시지를 rawValue에 추가해주지 않아도 빌드 상으로는 잘 돌아갑니다. 이런 점을 좀 피하고 싶어서 프로퍼티로 enum으로 제약?을 두고 싶었습니다. (case를 다 구현하지 않으면 빌드되지 않으니까요) 혹시 도미닉은 다른 의견 있을까요?!

AppleCEO commented:
- guard let을 썼던 이유는 옵셔널 바인딩을 위해 썼었습니다. 와일드 카드 패턴을 이용하지 않고 확인하는 방법으로는guard self.image != nil이 생각나는데 이 부분을 말씀하신걸까요?

네 그런 방식으로 비교하지 않은 이유가 있는지 궁금했습니다.

- Error에 관한 enum여서 Error 메시지는 반드시 설정해야 한다고 생각했습니다. rawValue에 지정하게 되면 Error가 추가되는 상황에서 메시지를 rawValue에 추가해주지 않아도 빌드 상으로는 잘 돌아갑니다. 이런 점을 좀 피하고 싶어서 프로퍼티로 enum으로 제약?을 두고 싶었습니다. (case를 다 구현하지 않으면 빌드되지 않으니까요) 혹시 도미닉은 다른 의견 있을까요?!

네 좋은 방식이라고 생각해서 질문했습니다. 👍 
연관되는 세미나 발표자료가 있는데 공유할께요~

AppleCEO commented:
참고하세요~ 
다음 스탭도 화이팅입니다!
https://youtu.be/yL0mPQskI4U?t=789
